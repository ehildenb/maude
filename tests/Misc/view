#!/bin/sh

MAUDE_LIB=$srcdir/../../src/Main
export MAUDE_LIB

../../src/Main/maude \
  < $srcdir/view.maude -no-banner -no-advise \
  > view.out 2>&1

diff $srcdir/view.expected view.out > /dev/null 2>&1
