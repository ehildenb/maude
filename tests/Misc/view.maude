set include BOOL off .

fth T is
  sort Elt .
  op f : Elt -> Elt .
endfth

fmod M{X :: T} is
  op g : X$Elt -> X$Elt .
var X : X$Elt .
  eq g(X) = f(f(X)) .
endfm

fmod FOO is
  sort Foo .
endfm

view V from T to FOO is
  sort Elt to Foo .
var X : Elt .
  op f(X) to term X .
endv

fmod TEST is
  inc M{V} .
endfm

show all TEST .

*** now invalidate view V by replacing FOO and recheck TEST

fmod FOO is
  sort Foo .
  op c : -> Foo .
endfm

show all TEST .

*** now replace view V altogether

view V from T to FOO is
  sort Elt to Foo .
var X : Elt .
  op f(X) to term c .
endv

show all TEST .

*** now replace T with incompatible theory and redefine everything

fth T is
  sort Elt .
  op f : Elt Elt -> Elt .
endfth

fmod M{X :: T} is
  op g : X$Elt -> X$Elt .
var X : X$Elt .
  eq g(X) = f(X, f(X, X)) .
endfm

fmod FOO is
  sort Foo .
  op h : Foo -> Foo .
endfm

view V from T to FOO is
  sort Elt to Foo .
vars X Y : Elt .
  op f(X, Y) to term h(Y) .
endv

fmod TEST is
  inc M{V} .
endfm

show all TEST .
