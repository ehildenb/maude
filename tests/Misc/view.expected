fmod TEST is
  sort Foo .
  op g : Foo -> Foo .
  eq g(X:Foo) = X:Foo .
endfm
fmod TEST is
  sort Foo .
  op c : -> Foo .
  op g : Foo -> Foo .
  eq g(X:Foo) = X:Foo .
endfm
fmod TEST is
  sort Foo .
  op c : -> Foo .
  op g : Foo -> Foo .
  eq g(X:Foo) = c .
endfm
fmod TEST is
  sort Foo .
  op h : Foo -> Foo .
  op g : Foo -> Foo .
  eq g(X:Foo) = h(h(X:Foo)) .
endfm
Bye.
